core/src/main/java/org/apache/calcite/util/format/postgresql/PostgresqlDateTimeFormatter.java

```diff
@@ -38,90 +45,164 @@ public class PostgresqlDateTimeFormatter {
+          0,
+          999_999,
+          6,
           dt -> String.format(Locale.ROOT, "%06d", dt.get(ChronoField.MICRO_OF_SECOND)),
           "US"),
       new NumberFormatPattern(
+          ChronoUnitEnum.TENTHS_OF_SECOND,
+          0,
+          9,
+          1,
           dt -> Integer.toString(dt.get(ChronoField.MILLI_OF_SECOND) / 100),
           "FF1"),
       new NumberFormatPattern(
+          ChronoUnitEnum.HUNDREDTHS_OF_SECOND,
+          0,
+          99,
+          2,
           dt -> String.format(Locale.ROOT, "%02d", dt.get(ChronoField.MILLI_OF_SECOND) / 10),
           "FF2"),
       new NumberFormatPattern(
+          ChronoUnitEnum.THOUSANDTHS_OF_SECOND,
+          0,
+          999,
+          3,
           dt -> String.format(Locale.ROOT, "%03d", dt.get(ChronoField.MILLI_OF_SECOND)),
           "FF3"),
       new NumberFormatPattern(
+          ChronoUnitEnum.TENTHS_OF_MS,
+          0,
+          9_999,
+          4,
           dt -> String.format(Locale.ROOT, "%04d", dt.get(ChronoField.MICRO_OF_SECOND) / 100),
           "FF4"),
       new NumberFormatPattern(
+          ChronoUnitEnum.HUNDREDTHS_OF_MS,
+          0,
+          99_999,
+          5,
           dt -> String.format(Locale.ROOT, "%05d", dt.get(ChronoField.MICRO_OF_SECOND) / 10),
           "FF5"),
       new NumberFormatPattern(
+          ChronoUnitEnum.THOUSANDTHS_OF_MS,
+          0,
+          999_999,
+          6,
           dt -> String.format(Locale.ROOT, "%06d", dt.get(ChronoField.MICRO_OF_SECOND)),
           "FF6"),
-      new EnumStringFormatPattern(ChronoField.AMPM_OF_DAY, "AM", "PM"),
-      new EnumStringFormatPattern(ChronoField.AMPM_OF_DAY, "am", "pm"),
-      new EnumStringFormatPattern(ChronoField.AMPM_OF_DAY, "A.M.", "P.M."),
-      new EnumStringFormatPattern(ChronoField.AMPM_OF_DAY, "a.m.", "p.m."),
-      new NumberFormatPattern(dt -> {
-        final String formattedYear = String.format(Locale.ROOT, "%0,4d", dt.getYear());
-        if (formattedYear.length() == 4 && formattedYear.charAt(0) == '0') {
-          return "0," + formattedYear.substring(1);
-        } else {
-          return formattedYear;
-        }
-      }, "Y,YYY") {
-        @Override protected String trimLeadingZeros(String value) {
-          return value;
-        }
-      },
-      new NumberFormatPattern(
+      new EnumStringFormatPattern(
+          ChronoUnitEnum.HALF_DAYS,
+          ChronoField.AMPM_OF_DAY,
+          "AM", "PM"),
+      new EnumStringFormatPattern(
+          ChronoUnitEnum.HALF_DAYS,
+          ChronoField.AMPM_OF_DAY,
+          "am", "pm"),
+      new EnumStringFormatPattern(
+          ChronoUnitEnum.HALF_DAYS,
+          ChronoField.AMPM_OF_DAY,
+          "A.M.", "P.M."),
+      new EnumStringFormatPattern(
+          ChronoUnitEnum.HALF_DAYS,
+          ChronoField.AMPM_OF_DAY,
+          "a.m.", "p.m."),
+      new YearWithCommasFormatPattern(),
+      new NumberFormatPattern(
+          ChronoUnitEnum.YEARS,
+          0,
```

Comment:
is 0 a legal value?

Reply:
PostgreSQL will accept `0` for year and silently convert it to 1 CE. I'll see if I have the behaviour right for this.

Reply:
Fixed up the behaviour here and added a test.